bjoern/server.c

```diff
@@ -117,20 +138,54 @@ pyerr_set_interrupt(struct ev_loop* mainloop, struct ev_cleanup* watcher, const
 static void
 ev_signal_on_sigint(struct ev_loop* mainloop, ev_signal* watcher, const int events)
 {
+#if defined(WANT_SIGNAL_HANDLING) || defined(WANT_GRACEFUL_SHUTDOWN)
+  ev_timer_stop(mainloop, &timeout_watcher);
+#endif
+
   /* Clean up and shut down this thread.
    * (Shuts down the Python interpreter if this is the main thread) */
   ev_cleanup* cleanup_watcher = malloc(sizeof(ev_cleanup));
   ev_cleanup_init(cleanup_watcher, pyerr_set_interrupt);
   ev_cleanup_start(mainloop, cleanup_watcher);
 
-  ev_io_stop(mainloop, &((ThreadInfo*)ev_userdata(mainloop))->accept_watcher);
+  ev_io_stop(mainloop, &THREAD_INFO->accept_watcher);
   ev_signal_stop(mainloop, watcher);
-#ifdef WANT_SIGNAL_HANDLING
-  ev_timer_stop(mainloop, &timeout_watcher);
+#ifdef WANT_GRACEFUL_SHUTDOWN
+  ev_signal_stop(mainloop, &THREAD_INFO->sigterm_watcher);
 #endif
 }
 #endif
 
+#ifdef WANT_GRACEFUL_SHUTDOWN
+static void
+ev_signal_on_sigterm(struct ev_loop* mainloop, ev_signal* watcher, const int events)
+{
+
+  ev_io_stop(mainloop, &THREAD_INFO->accept_watcher);
+  // Drain the accept queue before we close to try and minimize connection resets
+  ev_io_on_requests(mainloop, &THREAD_INFO->accept_watcher, 0);
+  // Close the socket now to ensure no more clients can talk to us
+  close(THREAD_INFO->server_info->sockfd);
+
+  ev_signal_stop(mainloop, watcher);
+  ev_timer_stop(mainloop, &timeout_watcher);
+
+  // don't shutdown immediately, but start a timer to check if we can shutdown in 100ms.
```

Comment:
Do we still need this if we change the counting to what I suggested above? It feels like this timeout stuff is unnecessarily complicated

Reply:
We would still need it, yes - without this there's nothing checking if there are any active connections, so it will always take 30 seconds to shutdown (i.e. the other timeout would be reached).

Reply:
Thinking a little bit more on this, I _could_ remove the `ev_shutdown_timeout_ontick` function and associated timer if I did the following:

- When SIGTERM is received, store `ev_now(loop) + 30` in an `ev_tstamp` on the server
- Make `ev_shutdown_ontick` check if `shutdown_time <= ev_now(loop)`, and use that to stop the loop.

That feels a bit simpler to me than timeout juggling it's currently doing.